picard/tagger.py

```diff
@@ -511,81 +526,60 @@ def add_files(self, filenames, target=None):
             log.debug("Adding files %r", new_files)
             new_files.sort(key=lambda x: x.filename)
             if target is None or target is self.unclustered_files:
-                self.unclustered_files.add_files(new_files)
                 target = None
             for file in new_files:
                 file.load(partial(self._file_loaded, target=target))
 
+    def _scan_dir(self, ignore_hidden, folders, recursive):
+        files = []
+        local_folders = list(folders)
+        while local_folders:
+            current_folder = local_folders.pop(0)
+            current_folder_files = []
+
+            try:
+                for entry in os.scandir(current_folder):
+                    if ignore_hidden and is_hidden(entry.path):
+                        continue
+                    if recursive and entry.is_dir():
+                        local_folders.append(entry.path)
+                    else:
+                        current_folder_files.append(entry.path)
+            except FileNotFoundError:
+                pass
+
+            number_of_files = len(current_folder_files)
+            if number_of_files:
+                mparms = {
+                    'count': number_of_files,
+                    'directory': current_folder,
+                }
+                log.debug("Adding %(count)d files from '%(directory)r'" %
+                          mparms)
+                self.window.set_statusbar_message(
+                    ngettext(
+                        "Adding %(count)d file from '%(directory)s' ...",
+                        "Adding %(count)d files from '%(directory)s' ...",
+                        number_of_files),
+                    mparms,
+                    translate=None,
+                    echo=None
+                )
+                files.extend(current_folder_files)
+        return files
+
     def add_directory(self, path):
-        if config.setting['recursively_add_files']:
-            self._add_directory_recursive(path)
-        else:
-            self._add_directory_non_recursive(path)
+        Thread(target=self._add_directory, args=[path, config.setting['recursively_add_files'], config.setting["ignore_hidden_files"]]).start()
 
-    def _add_directory_recursive(self, path):
-        ignore_hidden = config.setting["ignore_hidden_files"]
-        walk = os.walk(path)
+    def _add_directory(self, path, recursive, ignore_hidden):
+        if not os.path.exists(path) or not os.path.isdir(path):
+            return
```

Comment:
Instead of doing those tests, I'd rather let `_scandir()` raises exceptions, and catch them below. File can be modified between this line and first `scandir()`` call, so it may happen anyway (unlikely, but possible).

Reply:
OK